token/test/ZetaIdxUniversalToken.test.ts

```diff
@@ -0,0 +1,219 @@
+        }
+    });
+
+    describe("Initialization", function () {
+        it("should initialize with correct tokens and ratios", async function () {
+            for (let i = 0; i < testTokens.length; i++) {
+                const [token, ratio] = await indexToken.getTokenInfo(i);
+                expect(token).to.equal(testTokens[i].address);
+                expect(ratio).to.equal(RATIOS[i]);
+            }
+        });
+
+        it("should not allow initialization twice", async function () {
+            await expect(
+                indexToken.initializeIndex(
+                    testTokens.map(t => t.address),
+                    RATIOS
+                )
+            ).to.be.revertedWithCustomError(indexToken, "AlreadyInitialized");
+        });
+
+        it("should not allow initialization with invalid ratios", async function () {
+            // Deploy a new instance for this test
+            const ZetaIdxUniversalToken = await ethers.getContractFactory("ZetaIdxUniversalToken");
+            const newIndexToken = await upgrades.deployProxy(
+                ZetaIdxUniversalToken,
+                [
+                    owner.address, // initialOwner
+                    "ZetaIdx",    // name
+                    "ZIDX",       // symbol
+                    gateway,      // gatewayAddress
+                    GAS_LIMIT,    // gas
+                    uniswapRouter // uniswapRouterAddress
+                ],
+                {
+                    initializer: 'initialize',
+                }
+            ) as ZetaIdxUniversalToken;
+
+            const invalidRatios = [40, 30, 20]; // Sums to 90, not 100
+            await expect(
+                newIndexToken.initializeIndex(
+                    testTokens.map(t => t.address),
+                    invalidRatios
+                )
+            ).to.be.revertedWithCustomError(newIndexToken, "InvalidRatio");
+        });
+    });
+
+    describe("Wrap", function () {
+        it("should wrap tokens correctly", async function () {
+            const wrapAmount = ethers.utils.parseEther("100");
+            
+            // Check initial balance is 0
+            expect(await indexToken.balanceOf(user.address)).to.equal(0);
+            
+            // Check contract has no underlying tokens
+            for (const token of testTokens) {
+                expect(await token.balanceOf(indexToken.address)).to.equal(0);
+            }
+            
+            // Approve tokens
+            for (let i = 0; i < testTokens.length; i++) {
+                const tokenAmount = wrapAmount.mul(RATIOS[i]).div(100);
+                await testTokens[i].connect(user).approve(indexToken.address, tokenAmount);
+            }
+
+            // Wrap tokens
+            await indexToken.connect(user).wrap(wrapAmount);
+
+            // Check user received index tokens
+            expect(await indexToken.balanceOf(user.address)).to.equal(wrapAmount);
```

Comment:
Ok so the user will have to sign one approval for each component token then just one for wrap. And they could theoretically sign an infinite approval?